src/liballoc/borrow.rs

```diff
@@ -479,3 +479,17 @@ impl<'a> AddAssign<Cow<'a, str>> for Cow<'a, str> {
         }
     }
 }
+
+#[stable(feature = "cow_str_add_assign_char", since = "1.41.0")]
+impl AddAssign<char> for Cow<'_, str> {
+    fn add_assign(&mut self, rhs: char) {
+        if let Cow::Borrowed(lhs) = *self {
+            let mut s = String::with_capacity(lhs.len() + rhs.len_utf8());
+            // 4 bytes more since we know `rhs.len_utf8() <= 4`
+            s.vec.buf.amortized_new_size(s.capacity(), 4).expect("capacity overflow");
```

Comment:
This doesn't actually do anything, it just computes the new size. You need to call RawVec::reserve in some form to allocate the right new capacity.

https://github.com/rust-lang/rust/blob/e2fa9527d44522b900c739cc82faa734f877bcda/src/liballoc/raw_vec.rs#L449-L450

Reply:
Once we know the new size, shouldn't `reserve_exact` do just what we want? Right?

So after we have the calculated new size, a simple:

```rust
s.reserve_exact(additional);
```

Should do it?